python/ray/air/execution/resources/request.py

```diff
@@ -0,0 +1,199 @@
+
+    Returns: Dict containing all resources summed up.
+    """
+    resources = {}
+    for bundle in bundles:
+        for k, v in bundle.items():
+            resources[k] = resources.get(k, 0) + v
+    return resources
+
+
+@PublicAPI(stability="beta")
+class ResourceRequest:
+    """Request for resources.
+
+    This class is used to define a resource request. A resource request comprises one
+    or more bundles of resources and instructions on the scheduling behavior.
+
+    The resource request can be submitted to a resource manager, which will
+    schedule the resources. Depending on the resource backend, this may instruct
+    Ray to scale up (autoscaling).
+
+    Args:
+        bundles: A list of bundles which represent the resources requirements.
+        strategy: The scheduling strategy to acquire the bundles. This is only
+            relevant if the resources are requested as placement groups.
+
+         - "PACK": Packs Bundles into as few nodes as possible.
+         - "SPREAD": Places Bundles across distinct nodes as even as possible.
+         - "STRICT_PACK": Packs Bundles into one node. The group is
+           not allowed to span multiple nodes.
+         - "STRICT_SPREAD": Packs Bundles across distinct nodes.
+        *args: Passed to the call of ``placement_group()``, if applicable.
+        **kwargs: Passed to the call of ``placement_group()``, if applicable.
+
+    """
+
+    def __init__(
+        self,
+        bundles: List[Dict[str, Union[int, float]]],
+        strategy: str = "PACK",
+        *args,
+        **kwargs,
+    ):
+        if not bundles:
+            raise ValueError("Cannot initialize a ResourceRequest with zero bundles.")
+
+        self._bundles = [
+            {k: float(v) for k, v in bundle.items() if v != 0} for bundle in bundles
+        ]
+
+        if not self._bundles[0]:
+            # This is when trainable itself doesn't need resources.
+            self._head_bundle_is_empty = True
+            self._bundles.pop(0)
+
+            if not self._bundles:
+                raise ValueError(
+                    "Cannot initialize a ResourceRequest with an empty head "
+                    "and zero worker bundles."
+                )
+        else:
+            self._head_bundle_is_empty = False
+
+        self._strategy = strategy
+        self._args = args
+        self._kwargs = kwargs
+
+        self._hash = None
+        self._bound = None
+
+        self._bind()
+
+    @property
+    def head_bundle_is_empty(self):
+        """Returns True if head bundle is empty while child bundles
+        need resources.
+
+        This is considered an internal API within Tune.
+        """
+        return self._head_bundle_is_empty
+
+    @property
+    @DeveloperAPI
+    def head_cpus(self) -> float:
+        return 0.0 if self._head_bundle_is_empty else self._bundles[0].get("CPU", 0.0)
+
+    @property
+    @DeveloperAPI
+    def bundles(self) -> List[Dict[str, float]]:
```

Comment:
This is copied from PlacementGroupFactory, so I'd like to keep the current API (since it's public)

Reply:
Alternatively in the `PlacementGroupResourceRequest`.

Reply:
There is no PlacementGroupResourceRequest, in fact resource requests should be explicitly independent of the resource backend used.

I think the relationship is pretty clear, a `PlacementGroupFactory` is a `ResourceRequest` and thus inherits the full API. I would even argue that we should get rid of PGF and just use ResourceRequests (which is a better name anyway). I can't think of any functionality we would have in a PGF that we wouldn't want to have in a resource request. A PGF as the name suggests is just a factory wrapper around PGs. A resource request is explicitly documented to be compatible with the PG API. Thus they are essentially the same

Reply:
Actually this is the main thing that I really don't understand. I don't see how it's possible for the request to be independent of the backend, when the API exposes parameters that are only applicable to `PlacementGroupResourceManager`. 

Even if it is documented to have an API that is compatible with placement groups, if this is ultimately the public API then it seems strange that we must adhere to the placement group API for even non-placement group resource requests. 

Furthermore, who will decide what type of `ResourceManager` is actually used? Is this controlled by the user?

In my mind there should be some sort of hierarchy:

```
- ResourceRequest
- FixedResourceRequest(ResourceRequest)
- PlacementGroupResourceRequest(ResourceRequest)
```

1. If the user passes in a `PlacementGroupResourceRequest`, it will be handled by a `PlacementGroupResourceManager`. 
2. If the user passes in a `FixedResourceRequest`, it will be handled by a `FixedResourceManager`.
3. If the user passes in a `ResourceRequest`, then Tune(?) has the flexibility to decide which `ResourceManager` implementation to use.